CMakeLists.txt

```diff
@@ -36,7 +36,7 @@ list(APPEND CMAKE_MODULE_PATH
 include(pkg-utils)
 include(project-utils)
 
-
+option(snapshots "Build snapshot installers, CI option requires SNAPSHOT_VERSION variable during initial cmake call" OFF)
```

Comment:
This still does not explain what a snapshot installer actually is.
And tbh the note about the CI seems weird as well. To me it sounds like that's nothing the end-user has to worry about :eyes:

Reply:
Unless they are building it with CI, they probably don't. It's mainly for that purpose.

Reply:
So this option is effectively only meant to be used on the CI builds?

Could you please explain what this will actually do? Aka: In what regard does a snapshot installer differ from a regular one (and in how far do the build processes differ). Maybe if I understand this, I can make a better proposal for what we could put in the help message :)

Reply:
It adds a number between 1 and 65535 to the 4th position of the version number and adds the word `Snapshot` to the end of the installed product name. I'd like to discuss some more options so they can be finalized. We can also add info the file name, if needed.

Reply:
Ah so the only thing that changes is the name of the installer?
How does this relate to the `SNAPSHOT_VERSION` variable?

If all it does is changing the name of the installer, I'm actually not sure whether it is worth the hassle to create a CLI option for it. I mean if you need that, you could simply build the installer as usual and then rename it :eyes:

Reply:
`SNAPSHOT_VERSION` gets written to the 4th section of the version number which is part of the installer db. (i.e. `1.4.0.1`, etc...). This is needed to allow newer snapshots to upgrade older versions similar to a normal release. Otherwise the installer would always be the same version (i.e. 1.4.0) in the installer db and manual uninstalls would need to happen with each snapshot release instead of an upgrade. The variable would be passed by the CI during the first call of `cmake` with the generator. This allows a counter to exist so we can just increment that 4th number at a frequency we choose.

Reply:
Ah okay I see.
But doesn't this then conflict with the final stable release that will then only have a version of `1.4.0`? I would expect that the installer will consider `1.4.0.1` to be newer than `1.4.0` and thus it probably doesn't upgrade :thinking:

Reply:
We have to set the "build" part of the version for the final release as well.

Reply:
I think that is not optimal.
Couldn't we use e.g. negative numbers for the build version so that when leaving it out for the actual release the version is taken as higher (assuming a missing field is assumed to be zero)?

Reply:
Unfortunately not, it's an unsigned 16 bit field.

Reply:
Hm okay.

How is the version of the actual release managed? To my understanding the changes at hand only cause the build number to be included  if the snapshot option is used...

Reply:
It's currently derived from the tag's name, we have to edit `mumble-version.py` so that it sets the "build" part of the version.

I would do the same for the `.rc` files.

Reply:
I can just use the main `PROJECT_VERSION` number and add `.0` to the end of the version in the installer db from the installer files. Unless we are using a four digit version number normally.

Reply:
> Ah okay I see.
> But doesn't this then conflict with the final stable release that will then only have a version of `1.4.0`? I would expect that the installer will consider `1.4.0.1` to be newer than `1.4.0` and thus it probably doesn't upgrade

In theory, it would be best to use the current stable version with the incremented build number as a pre-release (i.e. beta, RC, snapshot) at least from the perspective of the version we are adding to the Windows installer. This is so that when the actual release happens the upgrades operate as expected and replace the previous version. We can add the prospective release number to the product or file name, if needed. So, pre-releases after `1.4.0` would appear normal. When `1.4.1` is being considered for release, all pre-release versions in the MSI files will be `1.4.0.<ci-build-count>`.  I can make another variable in the CMake project, possibly called `CURRENT_STABLE_VERSION` unless it would be a better idea to make the practice more consistent. I understand macOS and Linux might have different possibilities for versioning.

Back to the discussion about the added option and variables, I am open to suggestions for them and any descriptions. I will most likely revise the prefix `snapshot` to `msi-build` to hopefully provide more clarity. Most use cases will not require these settings, so we could also prefix them with `ci` if that helps.

Reply:
> In theory, it would be best to use the current stable version with the incremented build number as a pre-release

From a versioning point of view I agree. From a user's point of view who might see that version number appear in e.g. the Control Panel, this could be very confusing though :thinking: 
But I guess it is the best alternative we currently have :eyes: 

In terms of the cmake variable, I would probably use something like this:
```
option(snapshot-installer "Use a snapshot-versioning scheme internally when building an installer" OFF)
```

With regards to the needed environment variable, I think it might be an easier overall process if we dropped the need for an environment variable and use another cmake variable that can be set via command line instead. In fact we could even do this in a way that deprecates the need for the currently added option by simply checking whether that special variable is empty or not.
```
set(snapshot-build-number "0" CACHE STRING "<description>")
```

Reply:
> In terms of the cmake variable, I would probably use something like this:
> 
> ```
> option(snapshot-installer "Use a snapshot-versioning scheme internally when building an installer" OFF)
> ```
> 
> With regards to the needed environment variable, I think it might be an easier overall process if we dropped the need for an environment variable and use another cmake variable that can be set via command line instead. In fact we could even do this in a way that deprecates the need for the currently added option by simply checking whether that special variable is empty or not.
> 
> ```
> set(snapshot-build-number "0" CACHE STRING "<description>")
> ```
That should be fine so long as we document that as the process for all pre-release builds. Normal releases can just use `packaging=ON`. To lower the required options to the CI I also suggest setting `packaging=ON` when `snapshot-installer=ON`. I will add that to the root CMakeLists.txt along with another var for `stable-version`. I agree, it may be confusing if someone were to see the version number, but it's technically correct. It's similar to the Arch package versioning in that it's the stable version with changes on it even if we are calling it out internally as the projected release.